### ISSUE TITLE
"Failed to set up" integration when HA is restarted

### ISSUE THREAD

**Issue by ben-willmore** [asantaga/wiserHomeAssistantPlatform#317]
This was discussed on the forum thread, but now seems to happen reliably on restart for me. The integration is unavailable on restart, and the integrations page in HA says "Failed to set up" for the Wiser integration. I seem to be able to get it to work by going to the 3-dot menu for the integration, then Reload, then 3-dot menu again, System options, Update.

```
2022-10-25 20:34:35.405 ERROR (MainThread) [homeassistant.config_entries] Error setting up entry WiserHeat043089 for wiser
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 357, in async_setup
    result = await component.async_setup_entry(hass, self)
  File "/config/custom_components/wiser/__init__.py", line 60, in async_setup_entry
    await async_update_device_registry(hass, config_entry)
  File "/config/custom_components/wiser/__init__.py", line 80, in async_update_device_registry
    (CONNECTION_NETWORK_MAC, data.wiserhub.system.network.mac_address)
AttributeError: 'NoneType' object has no attribute 'network'
```

**Comment by msp1974** [comment]
Can you provide a diagnostic output please as have not been able to create this issue.

**Comment by thecem** [comment]
For me the same since the last update, beside the reload do not solve the problem. I only have a floorheating system with 4 Thermostats and the bridge version 1. Logs:

![A9849823-02FE-4D76-B9B2-A1CE6FCFBF02](https://user-images.githubusercontent.com/46648579/197952670-28972a47-d6d5-4556-8bed-75957b4e8c36.jpeg)


```
Dieser Fehler wurde von einer benutzerdefinierten Integration verursacht

Logger: homeassistant.config_entries
Source: custom_components/wiser/helpers.py:6 
Integration: Drayton Wiser Integration for Home Assistant (documentation, issues) 
First occurred: 25. Oktober 2022 um 09:03:50 (5 occurrences) 
Last logged: 08:33:22

Error setting up entry WiserHeat02E4DA for wiser
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 365, in async_setup
    result = await component.async_setup_entry(hass, self)
  File "/config/custom_components/wiser/__init__.py", line 60, in async_setup_entry
    await async_update_device_registry(hass, config_entry)
  File "/config/custom_components/wiser/__init__.py", line 82, in async_update_device_registry
    identifiers={(DOMAIN, get_identifier(data, 0))},
  File "/config/custom_components/wiser/helpers.py", line 69, in get_identifier
    return f"{data.wiserhub.system.name} {get_device_name(data, id, type)}"
  File "/config/custom_components/wiser/helpers.py", line 6, in get_device_name
    device = data.wiserhub.devices.get_by_id(id)
AttributeError: 'NoneType' object has no attribute 'get_by_id'

Logger: homeassistant.config_entries
Source: helpers/entity_component.py:160 
First occurred: 08:32:59 (32 occurrences) 
Last logged: 08:33:22

Error setting up entry WiserHeat02E4DA for select
Error setting up entry WiserHeat02E4DA for button
Error setting up entry WiserHeat02E4DA for number
Error setting up entry WiserHeat02E4DA for light
Error setting up entry WiserHeat02E4DA for cover
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 365, in async_setup
    result = await component.async_setup_entry(hass, self)
  File "/usr/src/homeassistant/homeassistant/components/climate/__init__.py", line 195, in async_setup_entry
    return await component.async_setup_entry(entry)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 160, in async_setup_entry
    raise ValueError("Config entry has already been setup!")
ValueError: Config entry has already been setup!


Logger: homeassistant.components.sensor
Source: custom_components/wiser/sensor.py:46 
Integration: Sensor (documentation, issues) 
First occurred: 25. Oktober 2022 um 09:03:50 (1 occurrences) 
Last logged: 25. Oktober 2022 um 09:03:50

Error while setting up wiser platform for sensor
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 281, in _async_setup_platform
    await asyncio.shield(task)
  File "/config/custom_components/wiser/sensor.py", line 46, in async_setup_entry
    hub_devices = data.wiserhub.devices.all
AttributeError: 'NoneType' object has no attribute 'all'

```

If you need more logs let me wich one and where.
Thx!

**Comment by ben-willmore** [comment]
Here are the diagnostics. I had to go through the reload-update cycle (so that the integration was functioning) before I had the option to download them. I hope that doesn't make them useless! Happy to provide any more info, or to dig around myself if helpful (I am competent with python etc but new to HA). Thanks for all your work!

[config_entry-wiser-2f19cc38495459ad1d8a320c7774cfbe.json-5.txt](https://github.com/asantaga/wiserHomeAssistantPlatform/files/9867021/config_entry-wiser-2f19cc38495459ad1d8a320c7774cfbe.json-5.txt)

**Comment by msp1974** [comment]
Ben, your diagnostic file was perfect.  I use this to emulate your hub in my mock hub server to make sure there is no data related issue that we do not catch.  This worked perfectly and loaded and functioned fine.

As such, I can only think that for some reason, it is failing getting an update from the hub at startup.  We have suppressed a lot of this but if you know how to pull a version from the dev branch and run that in your HA environment then, I can enable some logging and we can see better what is happening.  Let me know.

I also have a couple of questions if you can provide answers:

1) Did you run any of the 3.2.0beta versions or migrate from 3.1.7 to 3.2.0?
2) Have you tried removing and re-adding the integration?

**Comment by msp1974** [comment]
@thecem - can you also please answer the 2 questions above plus:

What version of HA are you running?
I presume you can not get it running at all on v3.2.0?  Is it ok if you go back to 3.1.7?
Can you provide a diagnostic output (from v3.1.7 or v3.2.0)

Thanks

**Comment by thecem** [comment]
HA Version:
![Uploading 524F922B-2BF5-41E5-89B0-4AD49037F314.jpeg…]()

**Comment by thecem** [comment]
![3687B8FA-34EC-40F5-9944-75E66F214DE3](https://user-images.githubusercontent.com/46648579/198019424-52051160-1e84-440f-988f-46b321a8532d.jpeg)

**Comment by thecem** [comment]
Going back to 3.1.7 is directly working.

**Comment by ben-willmore** [comment]
Yes I’m very happy to try a dev version. 

> * Did you run any of the 3.2.0beta versions or migrate from 3.1.7 to 3.2.0?

Yes, I ran most of the betas but not sure exactly which.

> * Have you tried removing and re-adding the integration?

No, I was afraid of breaking automations etc. but perhaps I don’t need to be?

**Comment by msp1974** [comment]
OK, Ben if you want to update form the dev branch and see if the logs show any better errors when you restart.

**Comment by thecem** [comment]
In 3.2.0 there is no diagnostic information button. In 3.1.7 it is:

[config_entry-wiser-71de748eec083b56f128f048f9ba6ada.json.txt](https://github.com/asantaga/wiserHomeAssistantPlatform/files/9869335/config_entry-wiser-71de748eec083b56f128f048f9ba6ada.json.txt)

**Comment by msp1974** [comment]
@thecem - ok this is a different issue.  Running your hub diagnostics on my emulator showed an issue with the UFH controller code.  I will transfer this to another issue and provide a fix in the next couple of days (once I know what this issue with Ben's is).

**Comment by ben-willmore** [comment]
I installed the dev branch, and couldn't reproduce the problem across 6 restarts. Instead, in the 'Integrations' page, I initially get "Retrying setup: None".  Then, after HA has finished starting up, this switches to the normal success card.

In the logs, I get:

```
$ grep wiser home-assistant.log
2022-10-26 18:17:48.869 WARNING (SyncWorker_0) [homeassistant.loader] We found a custom integration wiser which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2022-10-26 18:18:37.053 ERROR (MainThread) [custom_components.wiser.coordinator] Connection timeout trying to communicate with Wiser Hub 192.168.178.170.  Error is 
2022-10-26 18:18:37.068 WARNING (MainThread) [homeassistant.config_entries] Config entry 'WiserHeat043089' for wiser integration not ready yet: None; Retrying in background
```

I then reinstalled 3.2.0 from HACS and got "Failed to set up" again (which doesn't resolve itself) in one of two restarts.

TLDR: dev branch seems to fix the problem for me

**Comment by msp1974** [comment]
Hmm.  It looks like your hub does not respond to the initial request within 10s.  

I made a couple of changes to both improve logging but also fail the setup and retry if no data from the hub has been retrieved.  That is the integration not ready message.

Not sure why it does not update on first request.  I would have thought timeout this is plenty long enough as mine responds in a few hundred ms.  I'll think a bit more if there is anything that might cause this problem with a slow responding hub but for now, I'll push a v3.2.1 version from the dev branch so at least it works more reliably and fixes thecem's issue also.

**Comment by ben-willmore** [comment]
Could the slowness be on the HA end? I'm running on a Pi Zero which is usually fine but it does take a long time to start up.

**Comment by msp1974** [comment]
That could be the issue - we did have a different retry strategy in 3.1.7 but all 3.2beta versions were the same.

Anyway, v3.2.1 is now available to download via HACs which should now work although you may get the setup retry on each restart.

**Comment by ben-willmore** [comment]
Yes, v3.2.1 works great for me. Thanks for working on this!
